src/examples/zkapps/dex/dex-with-actions.ts

```diff
@@ -0,0 +1,367 @@
+
+class Dex extends SmartContract {
+  // addresses of token contracts are constants
+  tokenX = addresses.tokenX;
+  tokenY = addresses.tokenY;
+
+  /**
+   * state which keeps track of total lqXY supply -- this is needed to calculate what to return when redeeming liquidity
+   *
+   * total supply is zero initially; it increases when supplying liquidity and decreases when redeeming it
+   */
+  @state(UInt64) totalSupply = State<UInt64>();
+
+  /**
+   * redeeming liquidity is a 2-step process leveraging actions, to get past the account update limit
+   */
+  reducer = Reducer({ actionType: RedeemAction });
+
+  events = {
+    'supply-liquidity': Struct({ address: PublicKey, dx: UInt64, dy: UInt64 }),
+    'redeem-liquidity': Struct({ address: PublicKey, dl: UInt64 }),
+  };
+
+  /**
+   * Initialization. We set _all_ permissions to impossible except the ones we explicitly need.
+   */
+  init() {
+    super.init();
+    let proof = Permissions.proof();
+    this.account.permissions.set({
+      ...Permissions.allImpossible(),
+      access: proof,
+      editState: proof,
+      editActionState: proof,
+      send: proof,
+    });
+  }
+
+  @method createAccount() {
+    this.token.mint({ address: this.sender, amount: UInt64.from(0) });
+  }
+
+  /**
+   * Mint liquidity tokens in exchange for X and Y tokens
+   * @param dx input amount of X tokens
+   * @param dy input amount of Y tokens
+   * @return output amount of lqXY tokens
+   *
+   * This function fails if the X and Y token amounts don't match the current X/Y ratio in the pool.
+   * This can also be used if the pool is empty. In that case, there is no check on X/Y;
+   * instead, the input X and Y amounts determine the initial ratio.
+   */
+  @method supplyLiquidityBase(dx: UInt64, dy: UInt64): UInt64 {
+    let user = this.sender;
+    let tokenX = new TokenContract(this.tokenX);
+    let tokenY = new TokenContract(this.tokenY);
+
+    // get balances of X and Y token
+    let dexX = AccountUpdate.create(this.address, tokenX.token.id);
+    let x = dexX.account.balance.getAndAssertEquals();
+
+    let dexY = AccountUpdate.create(this.address, tokenY.token.id);
+    let y = dexY.account.balance.getAndAssertEquals();
+
+    // // assert dy === [dx * y/x], or x === 0
+    let isXZero = x.equals(UInt64.zero);
+    let xSafe = Circuit.if(isXZero, UInt64.one, x);
+    let isDyCorrect = dy.equals(dx.mul(y).div(xSafe));
+    isDyCorrect.or(isXZero).assertTrue();
+
+    tokenX.transfer(user, dexX, dx);
+    tokenY.transfer(user, dexY, dy);
+
+    // calculate liquidity token output simply as dl = dx + dx
+    // => maintains ratio x/l, y/l
+    let dl = dy.add(dx);
+    this.token.mint({ address: user, amount: dl });
+
+    // update l supply
+    let l = this.totalSupply.get();
+    this.totalSupply.assertEquals(l);
+    this.totalSupply.set(l.add(dl));
+
+    // emit event
+    let key = 'supply-liquidity' as const satisfies keyof Dex['events'];
```

Comment:
Interesting use of the typing here! I'm wondering if we chose this method because the event structs are defined inline or if it's due to concerns about type safety when emitting events and the inference isn't great here. Since this pattern appears a few times in the file, do you think it might be helpful to create a wrapper type for events that simplifies these statements for users to use if they run into this? I'd love to hear your thoughts on this!

Reply:
This complicated typing is not strictly necessary. What it does: it will cause a type error when changing the types in the `events` property on the smart contract. In the future, I want to enable such type safety by default via https://github.com/o1-labs/snarkyjs/issues/664, at which point we can remove this code